include/deal.II/lac/vector.templates.h

```diff
@@ -95,38 +94,58 @@ Vector<Number>::Vector (const Vector<OtherNumber> &v)
 
 
 #ifdef DEAL_II_WITH_PETSC
-
-template <typename Number>
-Vector<Number>::Vector (const PETScWrappers::Vector &v)
-  :
-  Subscriptor(),
-  vec_size(v.size()),
-  max_vec_size(v.size()),
-  val(nullptr)
+namespace internal
 {
-  if (vec_size != 0)
-    {
-      allocate();
-
-      // get a representation of the vector
-      // and copy it
-      PetscScalar *start_ptr;
-      PetscErrorCode ierr = VecGetArray (static_cast<const Vec &>(v), &start_ptr);
-      AssertThrow (ierr == 0, ExcPETScError(ierr));
-
-      internal::VectorOperations::copy (start_ptr, start_ptr+vec_size, begin());
-
-      // restore the representation of the
-      // vector
-      ierr = VecRestoreArray (static_cast<const Vec &>(v), &start_ptr);
-      AssertThrow (ierr == 0, ExcPETScError(ierr));
-    }
+  template <typename Number>
+  void
+  copy_petsc_vector(const PETScWrappers::VectorBase &v,
+                    ::dealii::Vector<Number> &out)
+  {
+    // do this in a two-stage process:
+    // first convert to a sequential PETSc
+    // vector, then copy that
+    Vec sequential_vector;
+    try
+      {
+        int ierr = VecCreateSeq (v.get_mpi_communicator(),
+                                 v.size(),
+                                 &sequential_vector);
+        AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+        ierr = VecCopy (v, sequential_vector);
+        AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+        PetscScalar *start_ptr;
+        ierr = VecGetArray(sequential_vector, &start_ptr);
+        AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+        const auto v_size = v.size();
+        if (out.size() != v_size)
+          out.reinit(v_size, true);
+
+        internal::VectorOperations::copy (start_ptr,
+                                          start_ptr + out.size(),
+                                          out.begin());
+        ierr = VecRestoreArray(sequential_vector, &start_ptr);
+        AssertThrow(ierr == 0, ExcPETScError(ierr));
+
+        ierr = VecDestroy (&sequential_vector);
+        AssertNothrow(ierr == 0, ExcPETScError(ierr));
```

Comment:
This is, by the way, something we can address with a custom deleter for a `std::unique` object -- see http://en.cppreference.com/w/cpp/memory/unique_ptr . In some sense, that would mean that we can attach a destructor to these types from the outside.